cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/batch/MapReduceContextConfigTest.java

```diff
@@ -51,23 +75,54 @@ public void testManyMacrosInAppSpec() {
+                                          Collections.<String, String>emptyMap(),
+                                          Collections.<String, DatasetCreationSpec>emptyMap(),
+                                          Collections.<String, FlowSpecification>emptyMap(),
+                                          Collections.<String, MapReduceSpecification>emptyMap(),
+                                          Collections.<String, SparkSpecification>emptyMap(),
+                                          Collections.<String, WorkflowSpecification>emptyMap(),
+                                          Collections.<String, ServiceSpecification>emptyMap(),
+                                          Collections.<String, ScheduleSpecification>emptyMap(),
+                                          Collections.<String, ScheduleCreationSpec>emptyMap(),
+                                          Collections.<String, WorkerSpecification>emptyMap(),
+                                          Collections.<String, Plugin>emptyMap());
     cfg.setApplicationSpecification(appSpec);
-    Assert.assertEquals(appSpec.getConfiguration(), cfg.getApplicationSpecification().getConfiguration());
+    Assert.assertEquals(appSpec.getConfiguration(),
+                        cfg.getApplicationSpecification().getConfiguration());
+  }
+
+  @Test
+  public void testGetPluginsWithMacrosMoreThan20() {
+    Configuration hConf = new Configuration();
+    MapReduceContextConfig cfg = new MapReduceContextConfig(hConf);
+    Map<String, Plugin> mockPlugins = new HashMap<String, Plugin>();
+    ArtifactId artifactId = new ArtifactId("plugins", new ArtifactVersion("1.0.0"), ArtifactScope.SYSTEM);
+    Map<String, String> properties = new HashMap<String, String>();
+    properties.put("path",
+                   "${input.directory}/${a}${b}${c}${d}${e}${f}${g}${h}${i}${j}"
+                     + "${k}${l}${m}${n}${o}${p}${q}${r}${s}${t}${u}${v}${w}${x}${y}${z}.txt");
+    hConf.set("input.directory", "/dummy/path");
+    String[] alphabetsArr = {"a", "b", "c", "d", "e", "f", "g", "h",
+      "i", "j", "k", "l", "m", "n", "o", "p", "q",
+      "r", "s", "t", "u", "v", "w", "x", "y", "z"};
+    for (String alphabet : alphabetsArr) {
+      hConf.set(alphabet, alphabet);
+    }
+    Set<ArtifactId> parents = new LinkedHashSet<>();
+    Plugin filePlugin1 = new Plugin(parents, artifactId,
+                                    new PluginClass("type", "name", "desc", "clsname", "cfgfield",
+                                                    Collections.<String, PluginPropertyField>emptyMap()),
+                                    PluginProperties.builder().addAll(properties).build());
+
+    mockPlugins.put("File1", filePlugin1);
+    hConf.set(HCONF_ATTR_PLUGINS, GSON.toJson(mockPlugins, PLUGIN_MAP_TYPE));
+
+    Map<String, Plugin> plugins = cfg.getPlugins();
+    Assert.assertEquals(plugins.size(), 1);
```

Comment:
should assert that the plugins are the same, not just the size.

Assert.assertEquals(mockPlugins, plugins)

Reply:
mockPlugins and plugins are not equal as PluginProperties build method is initializing macros with HashSet but the return from the GSON.fromJson is LinkedHashSet.

Reply:
Set equality does not depend on the Set implementation. Are you sure they are not different?

Reply:
In Plugin.java, only the macros equality check is failing.
&& Objects.equals(properties.getMacros(), that.properties.getMacros());
Created a simple test
Set<String> lh1 = new LinkedHashSet();
Set<String> h1 = new HashSet();
Macros m1 = new Macros(lh1);
Macros m2 = new Macros(h1);
System.out.println("Is m1 equal to m2: "+ m1.equals(m2));

Output
Is m1 equal to m2: false

Reply:
This has nothing to do with different set implementations. It is because the Macros class does not implement equals() and hashCode() on the release/4.3 branch. They are there on develop, you can copy-paste from there.